### ISSUE TITLE
An error in numeric assortativity coefficient for negative integers

### ISSUE THREAD

**Issue by vpozdnyakov** [networkx/networkx#4369]
### Environment Details

NetworkX version: 2.5.1
Python version: 3.7.10
Operating System: Google Colab web platform

### Error Description

Calculation numeric assortativity coefficient for negative integers returns a `nan` value.  A warning description is
```
/opt/anaconda3/lib/python3.7/site-packages/networkx/algorithms/assortativity/correlation.py:287: 
RuntimeWarning: invalid value encountered in double_scalars
  return (xy * (M - ab)).sum() / numpy.sqrt(vara * varb)
```
### Steps to reproduce

```python
import networkx as nx
G = nx.path_graph(2)
G.nodes[0]['attr'] = -2
G.nodes[1]['attr'] = -1
nx.numeric_assortativity_coefficient(G, 'attr')
```

### Expected behavior

Assortativity coefficient is -1.

Here is my reasoning. In the original Newman's article https://arxiv.org/abs/cond-mat/0209450, the numeric assortativity coefficient is defined as

<img src=https://user-images.githubusercontent.com/12996098/119843827-9380b600-bf10-11eb-8f52-ef36719649e9.png width=200>

where `x` and `y` are nodes' attribute values, `e_{xy}` is the fraction of links that follow from nodes of value `x` to one of value `y`, `a_x` is the fraction of links that follow *to* nodes of value `x` *from* all others, `b_y` is the fraction of links that follow *from* nodes of value `y` *to* all others, `\sigma_a` and `\sigma_b` are standard deviations of distributions `a_x` and `a_y`. The distribution `a_x` here means a discrete distribution with a support is all nodes' values, and the probability is the fraction of links *to* `x`. The distribution `b_y` has the same support as `a_x`, but the probability is the fraction of links *from* `y`. In case of an undirected network, the distributions `a_x` and `b_y` are equal. In this sense, my example can be calculated as follows.

**Step 1** Make a list of pairs of nodes's values `x` and `y`: (-1, -1), (-2, -1), (-1, -2), (-2, -2)
**Step 2** Calculate the fractions of edges `e_{xy}` for each pair: (-1, -1): 0, (-2, -1): 0.5, (-1, -2): 0.5, (-2, -2): 0
**Step 3** Calculate the fractions of edges `a_x` for each value `x`: (-1): 0.5, (-2): 0.5
**Step 4** The fractions of edges `b_y` for each value `y` are the same due to the undirected network.
**Step 5** Calculate the standard deviation `\sigma_a`. Let us use the formula 
<img src=https://user-images.githubusercontent.com/12996098/119851043-c2018f80-bf16-11eb-9483-8da8eee530db.png width=300> 

First, calculate the expectation `E[X] = (-1) * 0.5 + (-2) *0.5 = -1.5 `

Next, calculate the second moment `E[X^2] = (-1)^2 * 0.5 + (-2)^2 * 0.5 = 1 * 0.5 + 4 * 0.5 = 2.5`

Thus, the standard deviation is 
`\sigma_a = \sqrt{E[X^2] - E[X]^2} = \sqrt{2.5 - (-1.5)^2} = \sqrt{0.25} = 0.5`

**Step 6** The standard deviation `\sigma_b` is the same due to undirected network.
**Step 7** Finally, let us compute the assorativity coefficient. 

The nominator is the sum: `(-1) * (-1 ) * (0 - 0.5 * 0.5) + (-2) * (-1 ) * (0.5 - 0.5 * 0.5) + (-1) * (-2 ) * (0.5 - 0.5 * 0.5) + (-2) * (-2 ) * (0 - 0.5 * 0.5) = 1 * (-0.25) + 2 * 0.25 + 2 * 0.25 + 4 * (-0.25) = -0.25 + 0.5 + 0.5 - 1 = -0.25`

The denominator is `0.5 * 0.5 = 0.25`

Thus, the assortativity coefficient is `nomunator/denominator = -0.25 / 0.25 = -1`.

Here can be some mistakes or typos, but the obtained value is intuitively correct, since the assortativity coefficient is very close to the correlation and thereby -1 means that here are no links between equal values. Moreover, if we calculate the coefficient in the same network with positive values using NetworkX, we also get -1.

**Comment by dschult** [comment]
Yes.... an error occur in this case. Shouldn't it?
What is your idea? Your recent two issue descriptions aren't complete thoughts and could be more helpful when restated as such.

**Comment by vpozdnyakov** [comment]
@dschult I just updated the description, hope it helps to fix the issue. I also trying to understand the current behavior and suggest a solution.

**Comment by dschult** [comment]
Thanks for this -- and for your careful reading and the PR to fix things.
I'm still not sure what this issue is reporting though. Are you saying that the current setup does not produce the result -1? What does happen?  Can you give a short example code that demonstrates your example above? (maybe it is in the tests you added in your PR. I haven't looked closely yet.

**Comment by vpozdnyakov** [comment]
You are absolutely right. Updated description says two things: (1) according to the original paper, it is feasible to calculate the assortativity coefficient and it is -1, and (2) the current setup does not produce the result -1. There are no code examples or any suggestions for fixing this issue, only a step-by-step explanation why it is really unexpected behavior from a theoretical point of view. I have suggested a small change to fix it in my PR. I am not sure if new tests are needed. Let me know, if it is appropriate. In any case, I suggest discussing my corrections in the PR comments.
